### DISCUSSION TITLE
Flicker with HierarchicalMenu in Next.js 14.2.5

### DISCUSSION THREAD

**Discussion by sefai** [algolia/instantsearch#6287] [Q&A]
I am using HierarchicalMenu with SEO friendly URLs. After clicking any menu item, HierarchicalMenu and products are updated with a client-side query. Then it shows Menu and products without HierarchicalMenu applied for a brief moment and then reverts to normal state. I have made a slow motion video, you can clearly see the flicker. What could be the reason?

https://github.com/user-attachments/assets/d36103b5-2659-4fd4-989a-ee43b3006871

**Comment by Haroenv** [comment]
Are both widgets using the same attribute? Not sure what could cause this as it's hard to see in the video. Do you have a reproduction of this?

**Reply by sefai** [reply]
What do you mean by both widgets? I only provide a custom routing object, it's hard to reproduce on latency because of the way I create seo url's for each menu item, but it's something like below:

If I remove getLocation from the router object I get a blank page during the flicker, slow motion video is attached.


https://github.com/user-attachments/assets/cad66af2-e4e1-4fc6-8189-c7f49b1c4d1f



```
<InstantSearchNext<UiState, RouteState>
      searchClient={client}
      indexName="index_name"
      future={{
        persistHierarchicalRootCount: false,
        preserveSharedStateOnUnmount: true,
      }}
      routing={{
        router: {
          cleanUrlOnDispose: false,
          createURL({ qsModule, routeState, location }) {
            ...
          },
          parseURL({ qsModule, location }) {
            ...
          },
        },
        stateMapping: {
          stateToRoute(uiState) {
            const indexKeys = Object.keys(uiState);

            if (indexKeys) {
              const indexUiState = uiState[indexKeys[0]];

              return {
                q: indexUiState.query,
                category: indexUiState.hierarchicalMenu?.["hierarchy.lvl0"],
                brands: indexUiState.refinementList?.brand,
                page: indexUiState.page,
              };
            }

            return {
              q: undefined,
              category: undefined,
              brands: undefined,
              page: undefined,
            };
          },

          routeToState(routeState) {
            const hierarchicalMenu: { [key: string]: string[] } = {};
            const refinementList: { [key: string]: string[] } = {};

            if (routeState.category) {
              hierarchicalMenu["hierarchy.lvl0"] = routeState.category;
            }

            if (routeState.brands) {
              refinementList["brand"] = routeState.brands;
            }

            const state = {
              index_name: {
                query: routeState.q,
                hierarchicalMenu: hierarchicalMenu,
                refinementList: refinementList,
                page: routeState.page,
              },
            };

            return state;
          },
        },
      }}
    >
      <Configure
        analytics={false}
        hitsPerPage={48}
        distinct={true}
        filters={filters}
        attributesToRetrieve={[
          ...
        ]}
      />
      <SearchBox />
      <div className="flex flex-col w-full transition-opacity">
        <div className="flex">
          <div className="sticky top-0 h-screen w-2/5 min-w-2/5 max-w-2/5 mt-1 transition-[top] ease-in-out duration-200 delay-0">
            <HierarchicalMenu
              attributes={[
                "hierarchy.lvl0",
                "hierarchy.lvl1",
                "hierarchy.lvl2",
                "hierarchy.lvl3",
                "hierarchy.lvl4",
                "hierarchy.lvl5",
              ]}
              showParentLevel={true}
              transformItems={transformHierarchicalMenuItems}
            />
            <RefinementList
              attribute="brand"
              searchable
              searchablePlaceholder=".."
              showMore
              limit={5}
              showMoreLimit={15}
              transformItems={(items) => {
                return items.map((item) => {
                  ...
                });
              }}
            />
          </div>
          <Hits transformItems={transformProducts} />
        </div>
        <div></div>

        <Pagination
          ...
        />
      </div>
    </InstantSearchNext>
```

**Reply by sefai** [reply]
I tried to create one: https://codesandbox.io/p/devbox/zgmz9f?migrateFrom=wzqpzg

**Reply by sefai** [reply]
Hello @Haroenv, did you get a chance to look at the sandbox?

**Reply by sefai** [reply]
Any updates, please @Haroenv

**Reply by dhayab** [reply]
Hi, it seems there is an unnecessary server-side reload because `createURL()` matches the mapping of an actual Next.js route.

This might be solved by overriding the `push()` method from InstantSearch's router to [perform shallow routing](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api):

```ts
router: {
  cleanUrlOnDispose: true,
  createURL() { /* ... */ },
  parseURL() { /* ... */ },
  push(url) {
    if ((this as any).isDisposed) {
      return;
    }
    window.history.pushState({}, '', url);
  },
},
```

**Reply by sefai** [reply]
Thank you @dhayab , this resolves the flicker, but it also removes the Next.js route call. I indeed need some additional data from the server. What will be the way to achieve this, get some additional data from the server, then let instantsearch to perform client-side query and hydrate the results?

**Reply by aymeric-giraudet** [reply]
Hi @sefai sorry for the late reply.

You would have to remove the `history.pushState` solution.

I found a userland workaround as it's quite specific, but you can try this component : 
```tsx
function SaveLatestResultsOnUnmount() {
  const { results } = useInstantSearch();

  useEffect(() => {
    if (!results.__isArtificial) {
      const initialResults = {
        [results.index]: {
          results: results._rawResults,
          state: results._state,
        },
      };
      // @ts-ignore
      window[Symbol.for("InstantSearchInitialResults")] = initialResults;
    }
  }, [results]);

  return null;
}
```

You can use put it as a children of `InstantSearchNext`. This will override `initialResults` which is kinda hacky but it works.
However the image component will still flicker as it's being remounted, but this is a Next.js thing.

You could also make an API route for the other stuff you need to update and make manual client-side requests for them.

**Reply by sefai** [reply]
Thank you @aymeric-giraudet 

useLayoutEffect does not cause a flicker, maybe it's better to use in this case?

**Reply by aymeric-giraudet** [reply]
Yes @sefai, good idea !

I was thinking, you could probably also render Hierarchical Menu items as Next.js links, that way the transition will be made sooner without flickering and there shouldn't be a need for this hack either

**Reply by sefai** [reply]
I've already tried that, in fact I've tried many different things using Hierarchical Menu with both Next.js and Remix.

* using Links from both frameworks => all InstantSearch components lose their data, a completely blank page until client side algolia query finishes.
* awaiting fetch(from api routes) then calling onNavigate of HierarchicalListProps => no way to provide a feedback to the ui, page remains constant until fetch and algolia query finishes, then an immediate change on all components. From a UX perspective this doesn't look good and it is also against the navigation patterns of both frameworks.
* for remix and Link, returning serverState with my custom data => all InstantSearch components remain as it is, i can provide a feedback like changing opacity of hits until transition occurs,  but there is an unnecessary duplicate client side query which I couldn't get rid off. I suspect it is because of [this](https://github.com/algolia/instantsearch/blob/849190e41072ed0e0ec1311df69ab923adf22252/packages/instantsearch.js/src/lib/utils/hydrateSearchClient.ts#L79). I think returning serverState within a Remix client-side navigation is not handled as in SSR case, so cache misses.
* lastly this issue => using your kinda hacky solution, useLayoutEffect and loading="eager" on images prevents flicker. But refinementLists are also making a server request which I suspect is a problem within Next.js and maybe Next15 can help but it requires React 19 and algolia does not support it yet.

There are really many different ways but every one of them has a caveat. So I would really appreciate an official way of getting server data before client side navigations whenever necessary using InstantSearch.

**Reply by sefai** [reply]
FYI
* for remix and Link => this is because of singleFetch. If I disable singleFetch this works.
